docs: Add ADR for scanner metadata writeback (#48)
open-component-model/open-delivery-gear: docs/adr/001-scanner-metadata-writeback.md

## Context and Problem Statement

ODG uses scanners like BlackDuck or ClamAV to identify security and compliance related findings for given OCM components.
Most of the scanners follow the same architecture, where one part of the system recognises well-known opensource software, and another one maps opensource software against a metadata database.

ODG treats these scanners as blackboxes, as it just puts in blob data and extracts metadata.
If the scanner-reported properties are faulty (e.g. a package is detected in a wrong version, or an incorrect license is reported), ODG has no means to influence the extracted metadata.

Allowing ODG to write back corrected metadata to scanners raises a conceptional question: **How should ODG collect/store scanner metadata writebacks and how should they be propagated to the scanners?**
Adding a general purpose back channel from ODG to scanners to simply push writebacks is suboptimal because:

* ODG is built around a plugin concept where scanners are exchangeable, the writebacks must live outside of the scanner
* Not every scanner supports metadata writebacks, custom logic is necessary depending on the worker

> * ODG intents to abstract away from the scanners, leaking scanner details to ODG users is contraproductive

Reviewer: ```suggestion
* ODG intents to abstract away from the scanners, leaking scanner details to ODG users raises a conflict
```

#### Scanner Metadata Writeback Model

```yaml
artefact:
  component_name: <str>
  component_version: <str>
  artefact_kind: resource | source
  artefact:
    artefact_name: <str>
    artefact_version: <str>
    artefact_type: <str>
    artefact_extra_id: <object>
meta:
  datasource: delivery-dashboard
  type: meta/scanner_writeback
data:
  sub_type: <str>  # discriminator, e.g. package-version, license
  # remaining fields are fixed per sub_type (see sub-schemas below)
```

#### Sub-Schema: `package-version`

_Compatible with the existing OCM-label based package version overwrite mechanism._

```yaml
data:
  sub_type: package_version
  package_name: <str>
  package_version_from: <str> | null
  package_version_to: <str>
```

_Remarks: In case no `package_version_from` is specified, the `package_version_to` is to be applied to all detected package versions of `package_name`.
In case `package_version_from` is set, the writeback must only be applied to the specified version.
All remaining detected versions must remain unchanged._

#### Sub-Schema: `license`

```yaml
data:
  sub_type: license

>   package_name: <str>

Reviewer: what about optional package version?
